main.go

```diff
@@ -95,30 +111,72 @@ func main() {
 
 	ctrl.SetLogger(zap.Logger(enableDebugLog))
 
-	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
-		Scheme:             scheme,
-		MetricsBindAddress: metricsAddr,
-		LeaderElection:     enableLeaderElection,
-	})
-	if err != nil {
-		setupLog.Error(err, "unable to start manager")
-		os.Exit(1)
-	}
-
 	nsSlice := func(ns string) []string {
 		trimmed := strings.Trim(strings.TrimSpace(ns), "\"")
 		return strings.Split(trimmed, ",")
 	}
 
-	watchNs := make(map[string]bool)
-	if len(watchNamespaces) > 0 {
-		for _, ns := range nsSlice(watchNamespaces) {
-			watchNs[ns] = true
+	// If watchNamespaces is set ignore excludeNamespaces
+	if len(excludeNamespaces) > 0 && len(watchNamespaces) <= 0 {
+		logger.Info("setting restricted namespace list for controller")
+		clientset, err := kubernetes.NewForConfig(ctrl.GetConfigOrDie())
+		if err != nil {
+			logger.Error(err, "unable to get api client")
+		}
+		api := clientset.CoreV1()
+		namespaces, err := api.Namespaces().List(metav1.ListOptions{})
+		if err != nil {
+			logger.Error(err, "unable to get namespaces")
+		}
+		for _, namespace := range namespaces.Items {
+			if !stringInSlice(namespace.Name, nsSlice(excludeNamespaces)) {
+				namespaceList = append(namespaceList, namespace.Name)
+			}
+		}
+		nsList := strings.Join(namespaceList, ",")
+		logger.Info("watching namespaces: " + nsList)
+	} else {
+		if len(watchNamespaces) > 0 {
+			logger.Info("setting restricted namespace list for controller")
+			namespaceList = nsSlice(watchNamespaces)
+			// Remove any namespaces in excludeNamepsaces
+			if len(excludeNamespaces) > 0 {
+				for _, namespacea := range nsSlice(excludeNamespaces) {
+					for i, namespaceb := range namespaceList {
+						if namespacea == namespaceb {
+							namespaceList[i] = namespaceList[len(namespaceList)-1]
+							namespaceList[len(namespaceList)-1] = ""
+							namespaceList = namespaceList[:len(namespaceList)-1]
+						}
+					}
+				}
+			}
```

Comment:
maybe we can work with two slices and combine them using the following function:

```
func UniqueSlice(input []string) []string {
	u := make([]string, 0, len(input))
	m := make(map[string]bool)

	for _, val := range input {
		if _, ok := m[val]; !ok {
			m[val] = true
			u = append(u, val)
		}
	}
	return u
}
```

you can call the function like:
`UniqueSlice(append(namespaceSlice, excludedNamespaces...))`

in that way we could encapsulate this better and do it with less code

if we do this in that way we don't need to check lengths of structures because the combination of empty structures will give as another empty one so in the else code block we could just have:

`
namespacesList := UniqueSlice(nsSlice(excludedNamespaces), namespacesList...)`

with this we could also avoid so much indentation :)

what do you think?

Reply:
If I understand this correctly this isn't quite the intention of the if/else big loop thing. This would merge the 2 arrays into one unique list of namespaces instead of removing namespaces from the slice that exist in the exclude namespaces.

I have re-written the code now and will create a new comment detailing what I have done.